### ISSUE TITLE
VSCode: Creating first Q# project fails

### ISSUE THREAD

**Issue by kuzminrobin** [microsoft/qsharp-compiler#848]
On [this page](https://docs.microsoft.com/en-us/quantum/quickstarts/install-command-line?tabs=tabid-vscode), [Develop with Q#](https://docs.microsoft.com/en-us/quantum/quickstarts/install-command-line?tabs=tabid-vscode#develop-with-q) section, "VS Code" tab, the step **Q#: Create New Project** fails. Error message:
> Command 'Q#: Create new project...' resulted in an error (Running the contributed command: 'quantum.newProject' failed.)

Noticed on Sunday 2021.01.31, it was working in the beginning of the week (2021.01.(25 - 27)).
Reinstalling the VS Code and the extension "Microsoft Quantum Development Kit for Visual Studio Code" _does not help_.

**System information**
Windows 10
**VS Code**
Version: 1.52.1 (user setup)
Commit: ea3859d4ba2f3e577a159bc91e3074c5d85c0523
Date: 2020-12-16T16:34:46.910Z
Electron: 9.3.5
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Windows_NT x64 10.0.19042

**Screenshots**
![image](https://user-images.githubusercontent.com/9372582/106396254-b3b5c380-63bb-11eb-9310-75aa8299519f.png)

**Comment by kuzminrobin** [comment]
Looks like `npm` is not installed (or got removed) on my machine, which can be _my bad_.
(I don't remember if any docs/tutorials were asking me to install `npm` (Node.js?) during last 7 days when I started touching Quantum)

**Log (Extension Host):**
```
[2021-01-31 13:20:32.135] [exthost] [info] extension host started
[2021-01-31 13:20:32.166] [exthost] [info] ExtensionService#_doActivateExtension vscode.microsoft-authentication {"startup":false,"extensionId":{"value":"vscode.microsoft-authentication","_lower":"vscode.microsoft-authentication"},"activationEvent":"onAuthenticationRequest:microsoft"}
[2021-01-31 13:20:32.167] [exthost] [info] ExtensionService#loadCommonJSModule file:///c:/Users/rokuzmin/AppData/Local/Programs/Microsoft VS Code/resources/app/extensions/microsoft-authentication/dist/extension.js
[2021-01-31 13:20:32.177] [exthost] [info] ExtensionService#_doActivateExtension vscode.debug-auto-launch {"startup":true,"extensionId":{"value":"vscode.debug-auto-launch","_lower":"vscode.debug-auto-launch"},"activationEvent":"*"}
[2021-01-31 13:20:32.177] [exthost] [info] ExtensionService#loadCommonJSModule file:///c:/Users/rokuzmin/AppData/Local/Programs/Microsoft VS Code/resources/app/extensions/debug-auto-launch/dist/extension
[2021-01-31 13:20:32.181] [exthost] [info] ExtensionService#_doActivateExtension vscode.emmet {"startup":true,"extensionId":{"value":"vscode.emmet","_lower":"vscode.emmet"},"activationEvent":"*"}
[2021-01-31 13:20:32.181] [exthost] [info] ExtensionService#loadCommonJSModule file:///c:/Users/rokuzmin/AppData/Local/Programs/Microsoft VS Code/resources/app/extensions/emmet/dist/node/emmetNodeMain
[2021-01-31 13:20:32.191] [exthost] [info] ExtensionService#_doActivateExtension vscode.git {"startup":true,"extensionId":{"value":"vscode.github","_lower":"vscode.github"},"activationEvent":"*"}
[2021-01-31 13:20:32.191] [exthost] [info] ExtensionService#loadCommonJSModule file:///c:/Users/rokuzmin/AppData/Local/Programs/Microsoft VS Code/resources/app/extensions/git/dist/main
[2021-01-31 13:20:32.205] [exthost] [info] ExtensionService#_doActivateExtension vscode.merge-conflict {"startup":true,"extensionId":{"value":"vscode.merge-conflict","_lower":"vscode.merge-conflict"},"activationEvent":"*"}
[2021-01-31 13:20:32.205] [exthost] [info] ExtensionService#loadCommonJSModule file:///c:/Users/rokuzmin/AppData/Local/Programs/Microsoft VS Code/resources/app/extensions/merge-conflict/dist/mergeConflictMain
[2021-01-31 13:20:32.209] [exthost] [info] ExtensionService#_doActivateExtension vscode.search-result {"startup":true,"extensionId":{"value":"vscode.search-result","_lower":"vscode.search-result"},"activationEvent":"*"}
[2021-01-31 13:20:32.209] [exthost] [info] ExtensionService#loadCommonJSModule file:///c:/Users/rokuzmin/AppData/Local/Programs/Microsoft VS Code/resources/app/extensions/search-result/dist/extension.js
[2021-01-31 13:20:32.280] [exthost] [info] ExtensionService#_doActivateExtension ms-vscode.js-debug {"startup":false,"extensionId":{"value":"ms-vscode.js-debug","_lower":"ms-vscode.js-debug"},"activationEvent":"onCommand:extension.js-debug.clearAutoAttachVariables"}
[2021-01-31 13:20:32.280] [exthost] [info] ExtensionService#loadCommonJSModule file:///c:/Users/rokuzmin/AppData/Local/Programs/Microsoft VS Code/resources/app/extensions/ms-vscode.js-debug/src/extension.js
[2021-01-31 13:20:32.357] [exthost] [info] ExtensionService#_doActivateExtension vscode.github {"startup":true,"extensionId":{"value":"vscode.github","_lower":"vscode.github"},"activationEvent":"*"}
[2021-01-31 13:20:32.358] [exthost] [info] ExtensionService#loadCommonJSModule file:///c:/Users/rokuzmin/AppData/Local/Programs/Microsoft VS Code/resources/app/extensions/github/dist/extension.js
[2021-01-31 13:20:32.364] [exthost] [info] ExtensionService#_doActivateExtension vscode.github-authentication {"startup":false,"extensionId":{"value":"vscode.github-authentication","_lower":"vscode.github-authentication"},"activationEvent":"onAuthenticationRequest:github"}
[2021-01-31 13:20:32.364] [exthost] [info] ExtensionService#loadCommonJSModule file:///c:/Users/rokuzmin/AppData/Local/Programs/Microsoft VS Code/resources/app/extensions/github-authentication/dist/extension.js
[2021-01-31 13:20:32.375] [exthost] [info] eager extensions activated
[2021-01-31 13:20:32.382] [exthost] [info] ExtensionService#_doActivateExtension eamodio.gitlens {"startup":false,"extensionId":{"value":"eamodio.gitlens","_lower":"eamodio.gitlens"},"activationEvent":"onStartupFinished"}
[2021-01-31 13:20:32.382] [exthost] [info] ExtensionService#loadCommonJSModule file:///c:/Users/rokuzmin/.vscode/extensions/eamodio.gitlens-11.1.3/dist/gitlens
[2021-01-31 13:20:32.447] [exthost] [info] ExtensionService#_doActivateExtension ms-vscode-remote.remote-wsl {"startup":false,"extensionId":{"value":"ms-vscode-remote.remote-wsl","_lower":"ms-vscode-remote.remote-wsl"},"activationEvent":"onStartupFinished"}
[2021-01-31 13:20:32.447] [exthost] [info] ExtensionService#loadCommonJSModule file:///c:/Users/rokuzmin/.vscode/extensions/ms-vscode-remote.remote-wsl-0.52.0/dist/extension
[2021-01-31 13:20:55.004] [exthost] [info] ExtensionService#_doActivateExtension quantum.quantum-devkit-vscode {"startup":false,"extensionId":{"value":"quantum.quantum-devkit-vscode","_lower":"quantum.quantum-devkit-vscode"},"activationEvent":"onCommand:quantum.newProject"}
[2021-01-31 13:20:55.004] [exthost] [info] ExtensionService#loadCommonJSModule file:///c:/Users/rokuzmin/.vscode/extensions/quantum.quantum-devkit-vscode-0.15.2101126940/out/extension
[2021-01-31 13:21:05.570] [exthost] [error] Error: Command failed with exit code 1: npm root -g --prefix C:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\.yo-repository --loglevel error
'npm' is not recognized as an internal or external command,
operable program or batch file.
	at makeError (c:\Users\rokuzmin\.vscode\extensions\quantum.quantum-devkit-vscode-0.15.2101126940\node_modules\execa\lib\error.js:59:11)
	at Function.module.exports.sync (c:\Users\rokuzmin\.vscode\extensions\quantum.quantum-devkit-vscode-0.15.2101126940\node_modules\execa\index.js:188:17)
	at Object.runPackageManager (c:\Users\rokuzmin\.vscode\extensions\quantum.quantum-devkit-vscode-0.15.2101126940\node_modules\yeoman-environment\lib\util\repository.js:114:18)
	at Object.get nodeModulesPath [as nodeModulesPath] (c:\Users\rokuzmin\.vscode\extensions\quantum.quantum-devkit-vscode-0.15.2101126940\node_modules\yeoman-environment\lib\util\repository.js:40:36)
	at Function.assign (<anonymous>)
	at new Environment (c:\Users\rokuzmin\.vscode\extensions\quantum.quantum-devkit-vscode-0.15.2101126940\node_modules\yeoman-environment\lib\environment.js:257:30)
	at Function.createEnv (c:\Users\rokuzmin\.vscode\extensions\quantum.quantum-devkit-vscode-0.15.2101126940\node_modules\yeoman-environment\lib\environment.js:113:12)
	at Object.createNewProject (c:\Users\rokuzmin\.vscode\extensions\quantum.quantum-devkit-vscode-0.15.2101126940\out\commands.js:21:22)
	at c:\Users\rokuzmin\.vscode\extensions\quantum.quantum-devkit-vscode-0.15.2101126940\out\extension.js:64:24
	at c:\Users\rokuzmin\.vscode\extensions\quantum.quantum-devkit-vscode-0.15.2101126940\out\commands.js:16:9
	at _executeContributedCommand (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:701:220)
	at _.$executeContributedCommand (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:701:595)
	at m._doInvokeHandler (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:892:926)
	at m._invokeHandler (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:892:618)
	at m._receiveRequest (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:891:252)
	at m._receiveOneMessage (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:890:6)
	at c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:888:136
	at l.fire (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:46:801)
	at v.fire (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:233:787)
	at c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:1094:104
	at l.fire (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:46:801)
	at v.fire (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:233:787)
	at t.PersistentProtocol._receiveMessage (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:238:1004)
	at c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:235:927
	at l.fire (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:46:801)
	at p.acceptChunk (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:231:195)
	at c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:230:547
	at Socket.t (c:\Users\rokuzmin\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:240:822)
	at Socket.emit (events.js:223:5)
	at addChunk (_stream_readable.js:309:12)
	at readableAddChunk (_stream_readable.js:290:11)
	at Socket.Readable.push (_stream_readable.js:224:10)
	at Pipe.onStreamRead (internal/stream_base_commons.js:181:23) quantum.newProject
```
**Q# Language Extension log:** 
```
[Info  - 1:21:08 PM] workspace folder: (Null)
```
(just a single line)

**Comment by kuzminrobin** [comment]
Installing the npm according to the instructions [here](https://phoenixnap.com/kb/install-node-js-npm-on-windows) solved my original problem. However I allowed the installer to install the dependencies and it has downgraded my Visual Studio Installer from 2019 down to 2017 and failed then :-)  
The subsequent launch of the Visual Studio Installer has passed through some recovery steps and hopefully will work fine (keeping the fingers crossed).

**Comment by samarsha** [comment]
That's weird, because you shouldn't need npm installed to use the VS Code extension (only to build it).

**Comment by cgranade** [comment]
@samarsha: We got another report from @tvarga78 of `npm` being required to use project templates from the VS Code extension. I wonder if there was a regression in the VS Code extension that's causing that? +@ricardo-espinoza

**Comment by geduardo** [comment]
@cgranade @samarsha we have replicated this issue from several independent users on Windows 10. It's a major issue since it's blocking most new users from using the Docs or completing the Learn modules.

**Comment by samarsha** [comment]
Possible culprit is #806...

**Comment by rifat19331279** [comment]
<img width="910" alt="Capture" src="https://user-images.githubusercontent.com/69009277/106672478-f36bdf00-6564-11eb-898e-10f2afe91028.PNG">

**Comment by ricardo-espinoza** [comment]
Thanks for the report, @kuzminrobin. This is most likely a regression in our last two releases of the extension. For the time being, if installing `npm` works for you as a workaround, I'd recommend that. However, it's not intended to be a requirement. We'll investigate and address soon.

**Comment by cgranade** [comment]
Looking at the version of yeoman-generator that we depend on as of #806, there's a call to `Object.assign({}, repository)`. That call copies all properties of the variable repository over to a new object. That repository in turn has a property for the root node_module path, which it calls `npm` to find. That property is not actually used, but gets invoked anyway by the `Object.assign` call. In more recent versions of yeoman-generator, that call has been replaced by an explicit call to the constructor for `repository`, which then does not have the same bug: https://github.com/yeoman/environment/blob/c1b3b21f789be935c6bd0e145362b9162ef5235e/lib/environment.js#L253

My suspicion is that downgrading yeoman-environment to avoid the dependency on axios that was causing the vuln also introduced this regression, as per @samarsha's analysis. As per discussion at https://github.com/yeoman/environment/issues/246, we should now be able to restore a more recent version of yeoman-environment since the upstream dependency that brought in the vulnerable version of axios should now be fixed.

**Comment by samarsha** [comment]
Thanks for investigating, @cgranade. #806 was my PR, sorry about that. :( Requiring npm is really easy to miss since we would probably have it installed already...

**Comment by rifat19331279** [comment]
<img width="918" alt="Capture" src="https://user-images.githubusercontent.com/69009277/106881398-57d88c80-6692-11eb-90f1-5f405d6c3a92.PNG">

**Comment by ricardo-espinoza** [comment]
This is fixed in the recently released QDK version 0.15.2102.128318.
